### ISSUE TITLE
Unexpected Webpack2 `use` merge

### ISSUE THREAD

**Issue by bmayen** [survivejs/webpack-merge#63]
Given the following merge:
```
merge.smart({
  rules: [{
    test: /\.js$/,
    use: 'babel'
  }]
}, {
  rules: [{
    test: /\.js$/,
    use: 'coffee'
  }]
});
```
I would expect the following according to the what the Webpack1 examples seem to suggest in the README:
```
{
  rules: [{
    test: /\.js$/,
    use: 'coffee'
  }]
}
```
Instead I get:
```
{
  rules: [{
    test: /\.js$/,
    use: ['babel','coffee']
  }]
}
```
There doesn't seem to be a way to overwrite a string value in `use` as it always gets converted to an array and merged.

**Comment by bebraw** [comment]
Yeah, good call. I'll add a test based on your excellent example and get this fixed. 👍

**Comment by bebraw** [comment]
Even though the fix itself is simple, I realized this leads to more questions:

* What should happen if the first `use` is a string and the latter is an array? Merging those as an array would seem sensible though replacing would feel valid too.
* What should happen if both are arrays? Same idea here. This would behave like `loaders`.

It's that mixed case which is the tricky one. It's a little problematic `use` supports multiple types.

**Comment by bebraw** [comment]
Here's an initial fix if you want to try it out btw:

```diff
diff --git a/src/join-arrays-smart.js b/src/join-arrays-smart.js
index 4934c50..5fd2eb8 100644
--- a/src/join-arrays-smart.js
+++ b/src/join-arrays-smart.js
@@ -39,6 +39,15 @@ function uniteRules(newRule, rule, strategy) {
     if (optionsKey) {
       rule[optionsKey] = newRule[optionsKey];
     }
+  } else if (newRule.use) { // newRule.use should override too
+    const optionsKey = newRule.options ? 'options' : newRule.query && 'query';
+
+    delete rule.loaders;
+    rule.use = newRule.use;
+
+    if (optionsKey) {
+      rule[optionsKey] = newRule[optionsKey];
+    }
   } else if ((rule.use || rule.loaders || rule.loader) && (newRule.use || newRule.loaders)) {
     const expandEntry = loader => (
       typeof loader === 'string' ? { loader } : loader
@@ -71,12 +80,16 @@ function uniteRules(newRule, rule, strategy) {
     switch (strategy) {
       case 'prepend':
         rule[loadersKey] = [
-          ...differenceWith(newEntries, entries, uniteEntries),
+          ...differenceWith(
+            newEntries, entries, uniteEntries
+          ),
           ...entries
         ].map(unwrapEntry);
         break;
       default:
-        rule[loadersKey] = unionWith(entries, newEntries, uniteEntries).map(unwrapEntry);
+        rule[loadersKey] = unionWith(
+          entries, newEntries, uniteEntries
+        ).map(unwrapEntry);
     }
   }
 
diff --git a/tests/merge-smart-tests.js b/tests/merge-smart-tests.js
index babb8bb..46fd7b4 100644
--- a/tests/merge-smart-tests.js
+++ b/tests/merge-smart-tests.js
@@ -138,6 +138,24 @@ function mergeSmartTest(merge, loadersKey) {
     assert.deepEqual(merge(a, result), result);
   });
 
+  it('should override use string values with ' + loadersKey, function () {
+    const a = {};
+    a[loadersKey] = [{
+      test: /\.js$/,
+      use: 'a'
+    }];
+    const result = {};
+    result[loadersKey] = [{
+      test: /\.js$/,
+      use: 'b'
+    }, {
+      test: /\.css$/,
+      use: 'b'
+    }];
+
+    assert.deepEqual(merge(a, result), result);
+  });
+
   it('should append loaders with ' + loadersKey, function () {
     const a = {};
     a[loadersKey] = [{
```

It overrides `use` no matter what and that should be fixed next if we can figure out good answers to the questions above.

**Comment by bebraw** [comment]
There's one more case with `use` actually. It accepts an object. So mixed use with string/array/object should be thought about. It feels like that string/array case and probably goes through the same logic.

**Comment by bmayen** [comment]
I would love to be able to configure this behavior via options to `merge`. In my use case, I am using a base config which is maintained by the community and I have some company-specific overrides/additions merged in via separate config files. Ideally, I never edit the community-maintained configs because I frequently pull in the latest updates to those. But there are cases here where I want the opposite of what the default merge strategy does.

Could merge accept a config such as:
```
{ 
    strategies: {
        use: 'replace' // | 'force-array' | 'default'
    }
}

**Comment by bebraw** [comment]
Yeah. #64 feels related. At the very least I'll make it so that you can write something like `module.rules.use` to match and tweak the behavior.

Btw, I have API redesign going on at #61. It would make the implementation composable (easier to maintain, document, and use).

**Comment by bmayen** [comment]
Thanks for the work on this! I'll hopefully be able to try out the above fix today.

**Comment by bebraw** [comment]
I made a little fix. I think you should be able to use

```javascript
const output = merge.smartStrategy(
  {
    'module.rules.use': 'replace'
  })(common, prod);
```

now to achieve the behavior you described.

Let's leave this issue open for now, though, as this needs careful thought. It would be tempting to drop smart merging in 3.0 as it's strategies that are valuable, not folding.

**Comment by bmayen** [comment]
Just tested this with 2.5.0 using the approached shown in the previous comment. I am still seeing the previous behavior. Both `use` values are strings and they are getting merged into an array containing both.

**Comment by bebraw** [comment]
@bmayen Ok, I'll check that tomorrow. There's something else biting. `use` is definitely complex to handle.

**Comment by bebraw** [comment]
I pushed a fix for this. It's included in `2.6.0`. Please open new issues if you run into troubles with the behavior.

**Comment by bmayen** [comment]
Thank you! Will do.
